### ISSUE TITLE
Aborting test run with `force` true causes apparent hang

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#3276]
Found this in the GUI, see TestCentric/testcentric-gui#337.

There are a number of outstanding issues that may be related, but I decided to create this one anyway since it's specific to the `force` argument being true.

When we cancel with `force` false, everything works as expected. All ongoing tests and suites terminate and signal completion. The run is signaled as complete.

When `force` is true, the tests are aborted and signaled as complete but all ongoing suites are merely abandoned. No event is received for them so they appear to still be running. No run finished event is ever received.

Note that our existing tests around this all pass because the run is actually aborted. It just doesn't look that way to any runner that relies on event notifications, which is why I referred to an "apparent hang" in the title.

Existing issues, which may be related to this: #2947, #2282, #166 (five years old!). #166 is assigned to @oznetmaster, the others are unassigned.

**Comment by oznetmaster** [comment]
As @CharliePoole pointed out, I first discussed this years ago.

Forced stop actually worked correctly when we had a separate thread for each test instance.  It needed a few tweaks, perhaps.

It stopped working when the decision was made to stop using a different thread for each instance.

I pointed that out at the time, but the comments from others indicated that no one really thought that was an important consideration.

**Comment by CharliePoole** [comment]
Actually, all the tests __are__ being stopped and tearing is skipped as required. Problem is that no events are being received to record termination of the suites.

**Comment by CharliePoole** [comment]
Hmm... looking back at #166 that's the same behavior we are seeing now. Tests are cancelled but we get no notifications. I left a comment for @oznetmaster on #166 to please remove himself if he isn't actually working on it.

**Comment by CharliePoole** [comment]
OK, I think I see one source of the problem. `WorkItem`'s Cancel method is cancelling the thread on which Cancel is called. The intent is clearly that it would be cancelling the thread on which it is executing, but the method is now called from various places, on various threads. Cancelling the thread of your caller can't be good! :cry: We want to cancel the threads executing tests!

This explains why, after the initial tests are cancelled, additional code that is intended to clean up after the threads runs into the `ThreadAbortException`. We need to explicily abort the threads of the executing workers, not whatever thread happens to make the request.

**Comment by CharliePoole** [comment]
I noticed that I self-assigned this in error last year, while I was traveling in Europe!

**Comment by CharliePoole** [comment]
@nunit/framework-team 

This issue has come up again in TestCentric/testcentric-engine#39.

Currently, the framework uses conditional compilation to simply ignore `force: true` on a `StopRun` call. I can see three theoretical options...

1. Actually force the process to end
2. Advise the caller that the forced stop isn't supported
3. Remove the boolean argument in .NET 5.0 or higher builds

These options only apply to a __forced__ stop. I'm assuming that a cooperative stop will continue to function correctly.

Which way you go will influence what I do in both the TestCentric and NUnit engines, so even if it isn't implemented right away it will be helpful if you can decide on a direction.

**Comment by CharliePoole** [comment]
@rprouse Some indication of the planned direction on this issue would go a long  way  to moving the engine toward 4.0.

**Comment by CharliePoole** [comment]
If you remove Force Stop in 4.0, that does simplify the engine's job a bit. The driver for 4.0 would also no longer support it and the killing of an out of process test could be handled at the ProcessRunner level by killing the process.

**Comment by CharliePoole** [comment]
Whoops! I take back my last comment (May 9)!

`ForceStop(true)` was doing two things...
1. It killed the threads running tests. we know that can't be done any longer, at least for .NET Core and Standard.
2. It returned the results of the tests, which had been completed so far.

If we don't have some way to get (2) from the framework after a cancel, then there will be no partial results to save for most runners. Theoretically, the runner could construct a tree with the results by using all the progress reports sent from the framework, but that's a fair amount of work and duplicates what is already done in the framework.

So I think you have to either (1) accept `StopRun(true) and return what you can or (2) provide some other way to ask the framework for the tree of results so far.

**Comment by manfred-brands** [comment]
Aborting a test with force cannot be done in .net core as there no longer is a `Thread.Abort` method. 
The only thing we could do is Exit the process.

I think unit now silently ignores _force_ but I'm not sure.

**Comment by CharliePoole** [comment]
@manfred-brands @jt129254 

There is an implied engine issue here in addition to the unaddressed framework issue.

#### Framework

Alhough cancel with force doesn't  work under .net core, there is still a lot of code that deals with force == true, including the snippet cited by @jt129254 . So the call was never removed as @rprouse commented and is not entirely ignored either, although I believe it no longer attempts to actually cancel the thread. I may have missed something, but as far as I know, nothing further has changed in how `Cancel(true)` is processed since I created this issue almost six years ago.

#### Engine 

Since the framework API still exists and the framework still uses force==true to select some code branches, the engine still calls it. It now contains code to deal with the fact that `Cancel(true)` no longer actually works. That code kills the process and then generates fake completion events for all outstanding tests. It's a hack I created in TestCentric and then ported to the NUnit engine when I returned to work on it. I originally thought would be temporary, until the release of the the 4.0 framework. Unfortunately 4.0 continued to "support" `Cancel(true)` under .net core, so the hack is still there. I'm open to suggestions as to a better way to deal with this in the 4.0 engine. If we could think of a non-breaking change for 3.20, I'd  be in favor of that as well. I'll create an engine issue to discuss options.

**Comment by manfred-brands** [comment]
@CharliePoole Should we use this ticket to remove support for `Cancel(true)`, even for .NET Framework?
There are two way to do so:
1. Remove the overload, making this a compile time breaking change.
2. Throw a PlatformNotSupportedException. Same as Microsoft does for `Thread.Abort`.  As the call doesn't really work at the moment, that solves the hanging.

Even though I could add detection in the NUnit.Analyzer, I don't think it will be called from actual tests, only from engine implementations.

**Comment by CharliePoole** [comment]
@manfred-brands Those are both possibilities. A third is to make `Cancel(true)` cause a process termination, preferably after sending a notice to any runner. That's my favorite as a user and maintainer of some runners. The runner doesn't care how a cancellation happens, it just wants it to be done.

Although I can't work on the framework, I'm happy to review anything you decide to do.

**Comment by jnm2** [comment]
In .NET 7+, there is a near equivalent of Thread.Abort, but it can only abort code which has opted in to be able to be aborted, and it's not recommended for production usage (likely due to the process data corruption that is impossible to predict when disallowing code from stopping cooperatively): <https://learn.microsoft.com/en-us/dotnet/api/system.runtime.controlledexecution.run>

**Comment by CharliePoole** [comment]
@jnm2 What are your thoughts about killing the process as a last resort? Of course, that may involve data corruption as well but it does get rid of the process completely.

Up to now, I would have said it's the job of the test writer to ensure that the data is (1) not production data at all and (2) correctly set up for each run with any old copies destroyed.

**Comment by manfred-brands** [comment]
@jnm2 I have used the Thread.Abort equivalent in my timeout replacement branch. It works fine for Windows, but the Linux build hangs. I haven't had time to investigate.
If I get this to work than I would like to .net6.0 as target and have .net462 and .net8.0 both supporting timeout.

**Comment by manfred-brands** [comment]
> Provide tests with a cooperative shutdown signal 

We could provide this using the `TestContext.CancellationToken`.
I think we discussed before that we could check the `CancellationToken` in every `Assert` call.
That would solve it for those that use the NUnit API, but not for ones using [Shouldly](https://www.nuget.org/packages/Shouldly) 
nor would it solve really hanging test where they hang without doing any `Assert` call, e.g. badly programmed loop.

But we could make it part of a multi-prong approach:
1. Cancel the `TestContext.CancellationToken`.
2. If I can make the [ControlledExecution.Run](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.controlledexecution.run?view=net-9.0) work that would interrupt the current test. Allowing `TearDown` code to run. 
3. If after a some time the test still hangs, exit the current process.

**Comment by CharliePoole** [comment]
@manfred-brands  Not sure what you're saying about apps that don't use the API. How do they run tests then? Should nunit be responsible for apps that run tests without going through the API?

Use of the cancellation token seems to be part of cooperative cancellation, i.e. `Cancel(false)` rather than `Cancel(true)`. `ControlledExecution.Run` would be cooperative if you can make the TearDowns run but otherwise I'd call it forced. Exiting the process seems doubtful to me for two reasons: (1) The framework doesn't really know if its running in or out of process. (2) For out-of-process tests, the runner will have to check for termination and kill the agent anyway.

Multi-pronged makes sense though. From an engine/runner point of view, I think we have to assume that any test (or framework) can fail to terminate properly and act accordingly.

**Comment by manfred-brands** [comment]
> Not sure what you're saying about apps that don't use the API. How do they run tests then? Should nunit be responsible for apps that run tests without going through the API?

They use `Test` attributes and NUnit to start any tests, but don't do any `Assert`.
If I remember correctly `Shouldy` throws its own `Exception`
So if we check the `CancelationToken` in `Assert.That` it will not help them, but then that is their problem (Sorry @jnm2).

**Comment by OsirisTerje** [comment]
@manfred-brands @stevenaw  Related to #3248 .   @manfred-brands commented on this in Feb 25.  

I know we had some hangs due to the engine, which was fixed in a later version of the adapter (since it embeds the engine).  

Can we then close this issue ?

**Comment by CharliePoole** [comment]
@OsirisTerje I added a workaround in the engine, which I assumed I would get rid of when this was fixed. The workaround is in `MasterTestRunner`, and uses an instance of `WorkItemTracker` to maintain a list of suites that have started but not yet finished. I assumed that the framework would eventually do something similar and send a `TestSuiteFinished` event for any cancelled suites as well as test cases.

I haven't looked at the framework code for a long time, so I don't know how feasible this is. It would be good if somebody considered it and then let me know whether `WorkItemTracker` will need to be a permanent feature of the engine rather than a temporary fix.

The benefit of fixing this in the framework, of course, is that it would work whether the engine is used or not.

**Comment by manfred-brands** [comment]
We cannot abort badly behaving tests that have endless loops or some deadlock in user code anymore (and never could do so safely).

All the framework does is set the `TestExecutionContext.ExecutionStatus` to `AbortRequested`.
Which will only stop running any `TearDown` code.

The framework cannot do much more, unless `NUnitTestAssemblyRunner.StopRun` calls `Environment.Exit()`

I think VS runs test in an process separate from VS, so exiting should be fine.
If  I recalled correctly Nunit-Console also starts external runners.

**Comment by OsirisTerje** [comment]
Ok, then we just close this.

**Comment by CharliePoole** [comment]
@manfred-brands To be clear, this issue never asked for the force=true to work under .NET Core. It asks for it to work _correctly_ when it does work, i.e. under .NET Framework. That is, it should not _appear to hang_ when it has actually succeeded, which is what it does.

@OsirisTerje 
This seems more like a "Won't fix" than "somebody else's problem."

**Comment by OsirisTerje** [comment]
>  it should not appear to hang when it has actually succeeded, which is what it does.

I have not seen that lately, are you sure it does ?

SEP => Because the runners should not try to force a stop, and currently there is nothing we can do.  But, Wont Fix or Can't fix is fine too.

**Comment by manfred-brands** [comment]
@CharliePoole if the user (test centric) does a force stop, I would think that it could expect that it doesn't get events, but can assume itself that any events not received mean test not completed.

However that depends on the agreed behaviour of the API.
The XML doc for _force_ says: _if true, kill any test-running threads_

I compare it to killing a process, it also won't tell me about things it didn't do.

**Comment by OsirisTerje** [comment]
> The framework cannot do much more, unless NUnitTestAssemblyRunner.StopRun calls Environment.Exit()

The VS can run in two ways:  1. The old TestHost.exe (separate process yes).  2) MTP platform, where the testassembly itself runs as a separate process.

When we did have the hang issues earlier, we tried to do an Environment.Exit in the adapter, that didn't work as we expected when using MTP.  So I don't think the framework should attempt to do that.  

It is currently working fine wrt to stopping, that is, as we have documented https://docs.nunit.org/articles/nunit/writing-tests/attributes/cancelafter.html?q=cancel  .  We don't have any open issues now with complaints on this.  The latest engine and adapter fixes seems to have "cured" it, together with the CancelAfter attribute.

**Comment by CharliePoole** [comment]
Well, you did have one open issue with somebody complaining, but you just closed it. :-)

IMO, your work in the adapter highlights the fact that each runner has to do it differently, i.e. adapter vs engine.

Nevertheless, are you now saying I can remove the temporary fixes from V3 and V4 engines, since the cancellation token is intended to deal with the problem?

UPDATE: I have verified that both V3 and V4 build and pass all tests without the workaround included.

**Comment by OsirisTerje** [comment]
> Nevertheless, are you now saying I can remove the temporary fixes from V3 and V4 engines, since the cancellation token is intended to deal with the problem?

No, please don't do that.  I'm still nervous about the whole hang issue. 

And yes, I am "fighting" both the old TestHost and the new MTP, and no time to dig more into those two beasts.

> Well, you did have one open issue with somebody complaining, but you just closed it. :-)

This one ?

Do you mean we still have someone actually complaining about this now ?

**Comment by CharliePoole** [comment]
I'm the complainant. I filed the issue. I filed it seven years ago and nobody till now said "We've decided this is your problem." While I would not have been happy with that answer back then, it's really maddening to get it now since it takes now.

As you know, I was "only" a user when I created it, having left the organization. I'm still not part of the framework team, so basically a user as far as this issue goes. I've popped up here periodically over the years to ask about it and I've created untested temporary code to fix it in both TestCentric and, after I returned to the project, in the Engine. Nobody till now said "This isn't our problem."

Now that you've declared it "Somebody Else's Problem" I'll need to either remove it or add testing it to my backlog of work. The Console/Engine tests pass for both V3 and V4 without that temporary code, which, of course, only proves that it's not adequately tested. This comes at a time when I'm struggling to simplify what I need to do here rather than wanting new complications.

Not issuing `SuiteFinished` events to match each `SuiteStarted` event feels like a framework bug to me, both from looking at the code and from the docs. If it seems necessary to skip sending those events, then I think that should at least be documented, including a list of the circumstances when the finished events will not be received.

I hope that clarifies things from my point of view.

Charlie

PS: Reviewing the extensive discussion, I've noticed that almost everyone has commented on this issue as if it were titled "Aborting test run with `force` should work under .NET Core." I hope I've clarified that it was never about that but about what happens in the circumstances where `StopRun(true)` actually works.

**Comment by manfred-brands** [comment]
@CharliePoole

> Not issuing SuiteFinished events to match each SuiteStarted event feels like a framework bug to me, both from looking at the code and from the docs.

I checked the code and it calls `WorkItemComplete` which sends `TestFinished` in finally clause for tests.
The ParalleldWorkItemDispatched calls WorkItemCanceled for each active test.

There are test checking for events and tests for force stop, but not one testing events on stop (force or otherwise)
I'll see if I can add one and see where it does not send the expected 2nd half of the pair.

Are you ok with renaming the title as it isn't that the frameworks hangs, but that it doesn't send the finished events causing the runner to wait forever?

**Comment by CharliePoole** [comment]
@manfred I'm OK with a title change. I called it an "apparent hang" because
it's not the framework that's hanging but the process that is waiting to
get notifications. In my workaround code,  those notifications are created
without attempting to actually do the teardown.

My preference is fro the framework to generate "finished" events
corresponding to every "started" event that it has sent. A workable
alternative is a clear specification of when individual events are sent
_vs_ when they can be assumed. For example, if the framework can guarantee
that the "Finished" event for a test assembly implies that all contained
fixtures have also finished (or been cancelled) then any runner can work
with that. Otherwise, it feels like the logical assumption is that anything
that has not issued a finished event is still running.

One thing I had not mentioned... the temporary code I wrote is in
MasterTestRunner, which is not really the right place to put it. The code
really only applies if we are running NUnit tests, which means that any
special code should be in the NUnit driver if it's in the engine at all.

**Comment by manfred-brands** [comment]
@CharliePoole I have an nunit test confirming the behaviour.

It seems that even on .NET Framework a _force_ stop does nothing unless the `WorkItem` test runs on its own separate thread.
```
                    // Exit if not running on a separate thread
                    if (_thread is null)
                        return;
```

So the behaviour is correct in the sense that the test doesn't ever finish.

I then added `RequiresThread` on the class, that create a thread in the Fixture WorkItem, but not in the Test being cancelled meaning it doesn't cancel any tests.

Moving the attribute to the individual tests will actually stop the tests and consequently the events will be sent.

So it seems that `ForceStop` even on .NET Framework is very niche and only works in 1 situation.
Note that this has been since at least #1187 11 years ago where the method `Cancel` was added to `WorkItem`.

@OsirisTerje Can we not remove `ForceStop` completely as in 99% of the cases it doesn't work.

See the PR #5372 for the tests

**Comment by OsirisTerje** [comment]
> Can we not remove ForceStop completely as in 99% of the cases it doesn't work.

This was the recommendation earlier too, https://github.com/nunit/nunit/issues/3276#issuecomment-1121769267  

So yes, agree.

**Comment by CharliePoole** [comment]
That makes sense to me too. In which case, I'll create an issue to remove it from the V4 engine as well. I think that will reduce the possibility of confusion. We can discuss details on that issue.

**Comment by OsirisTerje** [comment]
So when #5374 is done, (no more force), this issue is done too, right ?

**Comment by manfred-brands** [comment]
There was never anything wrong with the events only with the assumption that force stop stopped all tests

**Comment by OsirisTerje** [comment]
Yes, I got that - but you agree we close this issue when #5374 is done?

We probably need to document this, could add something to the release notes, an Important section in this issue would do the thing, or in the other issue for the PR.

**Comment by CharliePoole** [comment]
@OsirisTerje 
Documenting this seems important for third-party runner authors.

Perhaps even more important from a user viewpoint... document how stopping works in any runner that exposes it. This wouldn't apply to the console runner but maybe to the adapter?

**Comment by OsirisTerje** [comment]
The adapter need to remove the StopRun when the engine is updated, but it relies on the work around in the engine.  

The user can cancel a run using Ctrl-C, both for MTP and non-MTP, e.g. cmd line dotnet.  
I need to work through that scenario, but it should just kill the processes, or let MTP kill itself. 
I'm not looking forward to that.........
